### ISSUE TITLE
directives rule

### ISSUE THREAD

**Issue by brettz9** [eslint/eslint#5505]
**When does this rule warn? Please describe and show example code:**

The proposed rule, `directives` would behave similarly to `strict`, except that it would allow behaviors to be defined for specific directives. Although I'm not sure whether all options as used with `strict` ("never", "global", "function", and "safe") would be necessarily useful, for parity, and to allow for flexibility (if not to also allow for merging the `strict` rule into this rule), I think it makes sense to allow the same options.

Here's a sample format:

```js
/*eslint directives: [2, {"use asm": "safe", "use babel": "global"}]
```

See https://github.com/eslint/eslint/pull/5451#issuecomment-192542560 for a further discussion of specific directives being used in the wild.

If you wish to merge this proposed rule with the `strict` rule, then one could also whitelist "use strict" (I use the full form since directives are not required to use "use ".)

**Is this rule preventing an error or is it stylistic?**

It is more in the middle, like a best practice, since it is most likely to be used for taking advantages of transformations such as Babel or optimization such as asm.js.

**Why is this rule a candidate for inclusion instead of creating a custom rule?**

Although the only spec-specified directive, `use strict` is already defined as a rule, there are a good number of directives out in the wild, and I think it makes sense to be able to have a uniform way of handling them.

**Are you willing to create the rule yourself?**

No.

**Comment by michaelficarra** [comment]
This definitely doesn't seem appropriate as a core rule. I doubt many eslint users have any directives at all in their programs, let alone the ones mentioned.

**Comment by brettz9** [comment]
Actually a big part of the rationale is to prevent unused strings in the directive position (besides `strict` and whitelisted ones) since these aren't covered by `no-unused-expressions`.

**Comment by ilyavolodin** [comment]
I'm really not sure how often directives other then `use strict` are used by majority of developers. It would be interesting to see how often other directives are being used. But I have a feeling that number is pretty low, and as such, would probably work better for a plugin, instead of being included into the core.

**Comment by brettz9** [comment]
What do you think of my other comment that the rule can also be used to catch useless string expressions which happen to occur in the prologue position by `/*eslint directives: 0*/`? Otherwise, they will not be caught (e.g., by `no-unused-expressions`).

**Comment by alberto** [comment]
Thanks for your interest in improving eslint. Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
